File: Search_Insert_Position.cpp
class Solution {
    /** 
     * param A : an integer sorted array
     * param target :  an integer to be inserted
     * return : an integer
     */
public:
    int searchInsert(vector<int> &A, int target) {
        // write your code here
        int s = 0;
        int e = A.size() - 1;
        while (s <= e){
            int m = s + (e - s) / 2;
            if (A[m] >= target){
                e = m - 1;
            }else {
                s = m + 1;
            }
        }
        return s;
    }
};

File: Search Insert Position.py
class Solution:
    """
    @param A : a list of integers
    @param target : an integer to be inserted
    @return : an integer
    """
    def searchInsert(self, A, target):
        # write your code here
        if not A:
            return 0
        
        left, right = 0, len(A) - 1
        while left + 1 < right:
            mid = left + (right - left) / 2
            # if A[mid] > target:
            if A[mid] >= target:
                right = mid
            elif A[mid] < target:
                left = mid
            else:
                return mid
        # if A[right] == target:
        #     return right
        # elif A[right] < target:
        #     return right + 1
        # if A[left] == target:
        #     return left
        # elif A[left] > target:
        #     return left
        # return left + 1
        if A[left] >= target:
            return left
        if A[right] < target:
            return len(A)
        else:
            return right